python/taichi/lang/matrix_ops_utils.py

```diff
@@ -36,7 +36,33 @@ def check(*args, **kwargs):
                 arg = args[i]
             except IndexError:
                 raise
-        do_check(fns, arg, **kwargs)
+        do_check(fns, arg)
+        return True, None
+
+    return check
+
+
+def foreach(*fns):
+    def check(args):
+        for x in args:
+            do_check(fns, x)
+        return True, None
+
+    return check
+
+
+def Or(f, g, msg=None):
+    def check(*args, **kwargs):
+        try:
+            do_check([f], *args, **kwargs)
+            return True, None
+        except TaichiCompilationError:
+            try:
```

Comment:
The other thing is `try-catch` has a maximum depth, and it's better only use `try-catch` in the outter-most function (`precondition` in your case). The inner functions should directly `raise` an Exception in some manner like: 
```
if xxxx:
    raise ...
```